vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,318 @@
 
-        // TODO: verify nonce of the from actor matches nonce of the message
+        let gas_cost = msg.gas_price() * msg.gas_limit();
+        // TODO requires network_tx_fee to be added as per the spec
+        let total_cost = &gas_cost + msg.value();
+        if from_act.balance < total_cost {
+            return Ok(ApplyRet::new(
+                MessageReceipt {
+                    return_data: Serialized::default(),
+                    exit_code: ExitCode::SysErrSenderStateInvalid,
+                    gas_used: BigUint::zero(),
+                },
+                miner_penalty_amount,
+                Some(ActorError::new(
+                    ExitCode::SysErrSenderStateInvalid,
+                    "Sender state invalid".to_owned(),
+                )),
+            ));
+        };
 
-        // TODO: check that the from actor has enough gas for the total gas cost
+        let mut gas_holder = ActorState::default(); // used for balance tracking
+        transfer_to_gas_holder(&mut self.state, msg.from(), &mut gas_holder, &gas_cost)?;
+        from_act.sequence += 1;
 
-        // TODO: transfer gas
+        let snapshot = self.state.snapshot()?;
 
-        // TODO: increase from actor nonce
+        // scoped to deal with mutable reference borrowing
+        let (ret_data, mut gas_used, act_err) = {
+            let (ret_data, mut rt, act_err) = self.send(msg, msg_gas_cost as i64);
+            rt.charge_gas(rt.price_list().on_chain_return_value(ret_data.len()))
+                .map_err(|e| e.to_string())?;
+            (ret_data, rt.gas_used(), act_err)
+        };
 
-        let return_data = self.send(msg, gas_cost.to_i64().unwrap());
-        return_data
-            .map(|r| {
-                Ok(MessageReceipt {
-                    return_data: r, // TODO: what about Send?,
-                    exit_code: ExitCode::Ok,
-                    gas_used: BigUint::from(0u64), // TODO: get from runtime, runtime.gas_used()
-                })
-            })
-            .map_err(|e| {
+        if let Some(err) = act_err {
+            if err.is_fatal() {
+                return Err(format!("Fatal send actor error occurred, err: {:?}", err));
+            };
+            if err.exit_code() != ExitCode::Ok {
+                // revert all state changes since snapshot
                 if let Err(state_err) = self.state.revert_to_snapshot(&snapshot) {
-                    return state_err;
-                }
-                e.to_string()
-            })?
+                    return Err(format!("Revert state failed: {}", state_err));
+                };
+            }
+            warn!("Send actor error: from:{}, to:{}", msg.from(), msg.to());
+        }
+        // TODO ask if this is correct as this could give a free tx if gas used was a negative value(e.g. delete_miner)
+        // see: https://github.com/filecoin-project/lotus/blob/testnet/3/chain/vm/vm.go#L388
+        if gas_used < 0 {
+            gas_used = 0;
+        }
```

Comment:
consider maybe something like
```suggestion
        let gas_used = if gas_used < 0 { 0 } else { gas_used as u64 };
```
so you don't have to do a bunch of conversions below